bin/server/cmd/component_omni.go

```diff
@@ -48,19 +54,70 @@ func createMetadataDB(activeProfile *Profile, logger *zap.Logger) (*metadataDB,
 }
 
 func (m *metadataDB) connect() (*store.DB, error) {
-	if err := m.pgInstance.Start(m.profile.datastorePort, os.Stderr, os.Stderr); err != nil {
+	if useEmbedDB() {
+		if err := m.pgInstance.Start(m.profile.datastorePort, os.Stderr, os.Stderr); err != nil {
+			return nil, err
+		}
+		m.pgStarted = true
+
+		// Even when Postgres opens Unix domain socket only for connection, it still requires a port as ID to differentiate different Postgres instances.
+		connCfg := dbdriver.ConnectionConfig{
+			Username:    m.profile.pgUser,
+			Password:    "",
+			Host:        common.GetPostgresSocketDir(),
+			Port:        fmt.Sprintf("%d", m.profile.datastorePort),
+			StrictUseDb: false,
+		}
+		db := store.NewDB(m.l, connCfg, m.profile.demoDataDir, readonly, version, m.profile.mode)
+		return db, nil
+	}
+
+	u, err := url.Parse(pgURL)
+	if err != nil {
 		return nil, err
 	}
-	m.pgStarted = true
 
-	// Even when Postgres opens Unix domain socket only for connection, it still requires a port as ID to differentiate different Postgres instances.
+	m.l.Info("Establishing external PostgreSQL connection...", zap.String("pgURL", u.Redacted()))
```

Comment:
Extract this method to root.go since it duplicates the code in slim?

Reply:
I thought about this, but I feel like `connecting()` located in xxx_slim/omni.go would be confusing to code readers if named like `connectExternal()`

Reply:
Extract a method like connectExternal to root.go and call that in slim/omni?

Reply:
Yes, it will very confused when you are coding at related logic, and your autocompletion prompt you there are a `connect()` and a `connectExternal()` :\

Reply:
Note, after we implement the code, we are unlikely to use write code to call connect/connectExternal in other places, so the prompt thing is less concern. 

The more concerning thing is for the reader to figure out the difference between connect and connectExternal. There are 2 ways to tackle this

1. create a separate component.go file and put the shared useEmbedDB() and connectExternal() there
2. Add a comment above connectExternal().

Duplication is less ideal here because like the other 2 bugs mentioned, we need to fix it in 2 places. And this is a commonly inspected code path for the --pg feature and there is quite a few "if else" logic, so keeping them in sync is more challenging than simple code duplication.

Reply:
Put it another way, when reading the logic for --pg, we don't want reader to read the logic in 2 places.

Reply:
> Note, after we implement the code, we are unlikely to write new code to call connect/connectExternal in other places, so the prompt thing is less of a concern.
> 
> The more concerning thing is for the reader to figure out the difference between connect and connectExternal. There are 2 ways to tackle this
> 
> 1. create a separate component.go file and put the shared useEmbedDB() and connectExternal() there
> 2. Add a comment above connectExternal().
> 
> Duplication is less ideal here because like the other 2 bugs mentioned, we need to fix it in 2 places. And this is a commonly inspected code path for the --pg feature and there is quite a few "if else" logic, so keeping them in sync is more challenging than simple code duplication.

Method 1 SG.